src/fetchcode/package_versions.py

```diff
@@ -0,0 +1,638 @@
+
+    Do a DFS for the `item` in the `dictionary` by traversing the `attributes`
+    and return None if can not traverse through the `attributes`
+    For example:
+    >>> get_item({'a': {'b': {'c': 'd'}}}, 'a', 'b', 'c')
+    'd'
+    >>> assert(get_item({'a': {'b': {'c': 'd'}}}, 'a', 'b', 'e')) == None
+    """
+    for attribute in attributes:
+        if not dictionary:
+            return
+        if not isinstance(dictionary, dict):
+            logger.error("dictionary must be of type `dict`")
+            return
+        if attribute not in dictionary:
+            logger.error(f"Missing attribute {attribute} in {dictionary}")
+            return
+        dictionary = dictionary[attribute]
+    return dictionary
+
+
+def cleaned_version(version):
+    """Return a ``version`` string stripped from leading "v" prefix."""
+    return version.lstrip("vV")
+
+
+def nuget_extract_versions(response: dict) -> Iterable[PackageVersion]:
+    for entry_group in response.get("items") or []:
+        for entry in entry_group.get("items") or []:
+            catalog_entry = entry.get("catalogEntry") or {}
+            version = catalog_entry.get("version")
+            if not version:
+                continue
+            release_date = catalog_entry.get("published")
+            if release_date:
+                release_date = dateparser.parse(release_date)
+            yield PackageVersion(
+                value=version,
+                release_date=release_date,
+            )
+
+
+def maven_extract_versions(
+    xml_response: ET.ElementTree,
+) -> Iterable[PackageVersion]:
+    for child in xml_response.getroot().iter():
+        if child.tag == "version" and child.text:
+            yield PackageVersion(value=child.text)
+
+
+def get_pypi_latest_date(downloads):
+    """
+    Return the latest date from a list of mapping of PyPI ``downloads`` or  None.
+
+    The data has this shape:
+    [
+      {
+        ....
+        "upload_time_iso_8601": "2010-12-23T05:14:23.509436Z",
+        "url": "https://files.pythonhosted.org/packages/8f/1f/c20ca80fa5df025cc/Django-1.1.3.tar.gz",
+      },
+      {
+        ....
+        "upload_time_iso_8601": "2010-12-23T05:20:23.509436Z",
+        "url": "https://files.pythonhosted.org/packages/8f/1f/561bddc20ca80fa5df025cc/Django-1.1.3.wheel",
+      },
+    ]
+    """
+    latest_date = None
+    for download in downloads:
+        upload_time = download.get("upload_time_iso_8601")
+        if upload_time:
+            current_date = dateparser.parse(upload_time)
+        if not latest_date:
+            latest_date = current_date
+        else:
+            if current_date > latest_date:
+                latest_date = current_date
+    return latest_date
+
+
+def get_response(url, content_type="json", headers=None):
+    """Fetch ``url`` and return its content as ``content_type`` which is one of binary, text or json."""
+    assert content_type in ("binary", "text", "json", "yaml")
```

Comment:
@JonoYang we can remove this assert altogether since this method is not meant to be directly accessed from outside and is just a utility function for different ecosystems to fetch the response in the appropriate format. The only way for this to go wrong is when the ecosystem function is incorrect in the first place.